lasso/emme.py

```diff
@@ -268,145 +268,190 @@ def create_emme_network(
         setup = SetupEmme(model_tables, out_dir, _NAME, include_transit, parameters)
         setup.run()
 
-def prepare_table_for_taz_drive_network(
-    nodes_df,
-    links_df,
-    parameters,
-):
-
-    """
-    prepare model table for taz-scale drive network, in which taz nodes are centroids
-    keep links that are drive_access == 1 and assignable
 
-    Arguments:
-        nodes_df -- node database
-        links_df -- link database
+def extract_bus_shapes(transit_network: StandardTransit, route_type_bus_id:int = 3):
+    """Returns a list of Nodes that contain bus routes"""
     
-    Return:
-        dictionary of model network settings
+    shapes_df = transit_network.feed["shapes"].copy()
+    trips_df = transit_network.feed["trips"].copy()
+    routes_df = transit_network.feed["routes"].copy()
+
+    bus_routes_df = routes_df.loc[routes_df["route_type"].isin([route_type_bus_id]), "route_id"]
+    bus_trips_df = trips_df.loc[trips_df["route_id"].isin(bus_routes_df), "shape_id"]
+    bus_shapes_df = shapes_df[shapes_df["shape_id"].isin(bus_trips_df)]
+    return bus_shapes_df
+
+def tag_if_link_contains_bus(links_df: gpd.GeoDataFrame, gtfs_shape_bus_routes: gpd.GeoDataFrame):
+    """tags nodes where the starting node and the end node are bus route nodes"""
+    links_df = links_df.copy()
+    gtfs_shape_bus_routes = gtfs_shape_bus_routes.copy()
+
+    # links are defined as pairs of nodes, so get next node (this assumes order)
+    gtfs_shape_bus_routes["next_node_id"] = gtfs_shape_bus_routes["shape_model_node_id"].shift(1)
```

Comment:
shift(1) is previous value, next value should be shift(-1). I think here you want the next value so shift(-1)

Reply:
We shift forward because we later drop the first link in a shapefile. Have renamed to previous node ID
See line
"gtfs_shape_bus_routes = gtfs_shape_bus_routes[(gtfs_shape_bus_routes["shape_pt_sequence"] != 1)]"

since dataframe looks like with this shift:
![image](https://github.com/user-attachments/assets/f9f5768a-a189-4d63-92b1-906e38e56fd1)